src/transfer.h

```diff
@@ -137,17 +145,44 @@ void transfer_to_particles(
     {
       vals[k] = f_array[dofs[k]];
     }
+    // Cast as matrix of size [block_size, space_dimension/block_size]
+    Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
+                                   Eigen::ColMajor>>
+        vals_mat(vals.data(), block_size, space_dimension);
+
     for (int pidx : cell_particles[c])
     {
       Eigen::Map<Eigen::VectorXd> ptr = field.data(pidx);
       ptr.setZero();
-      Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
-                                     Eigen::ColMajor>>
-          q(basis_values.row(idx++).data(), value_size, space_dimension);
+      Eigen::Map<const Eigen::VectorXd> q(basis_values.row(idx++).data(),
+                                          space_dimension);
 
-      ptr = q * vals;
+      ptr = vals_mat * q;
     }
   }
 }
+//----------------------------------------------------------------------------
+void eval_particle_cell_contributions(
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>& q,
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>& l,
+    int* row_idx, const Particles& pax, const Field& field, int cidx,
+    int space_dimension, int block_size,
+    const Eigen::Array<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>&
+        basis_values)
```

Comment:
This signature looks a bit messy - we've been trying to get away from [in/out] arguments in dolfinx, i.e. non-const by reference, and pointers. Maybe we could do a bit more outside this function, i.e. send in `cell_particles` as a const argument, and return `q` and `l` as a tuple.

Reply:
Also `row_idx` could just be a local variable, not a pointer.

Reply:
Updated the signature of `eval_particle_cell_contributions` a bit, now
- returning a `std::pair` with 2 `Eigen::MatrixXd` as entries
- taking a const `row_offset` rather than a `row_idx` pointer